src/coreComponents/functions/MultivariableTableFunctionKernels.hpp

```diff
@@ -0,0 +1,327 @@
+   * @param[in] axisPoints number of discretization points for a given axis
+   * @param[out] axisLow left coordinate of target axis interval
+   * @param[out] axisMult weight of the right coordinate of target axis interval
+   * @return integer target axis interval index
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  integer
+  getAxisIntervalIndexLowMult( real64 const axisCoordinate,
+                               real64 const axisMin,
+                               real64 const axisMax,
+                               real64 const axisStep,
+                               real64 const axisStepInv,
+                               integer const axisPoints,
+                               real64 & axisLow,
+                               real64 & axisMult ) const
+  {
+    integer axisIntervalIndex = integer((axisCoordinate - axisMin) * axisStepInv );
+
+    // check that axisindex is within interpolation interval: valid axisindex is between 0 and (axisPoints - 2),
+    // since there are axisPoints-1 intervals along the axis
+
+    if( axisIntervalIndex < 0 )
+    {
+      axisIntervalIndex = 0;
+      if( axisCoordinate < axisMin )
+      {
+        printf( "Interpolation warning: axis coordinate is out of limits (%lf; %lf) with value %lf, extrapolation is applied\n", axisMin, axisMax, axisCoordinate );
+      }
+    }
+    else if( axisIntervalIndex > (axisPoints - 2))
+    {
+      axisIntervalIndex = axisPoints - 2;
+      if( axisCoordinate > axisMax )
+      {
+        printf( "Interpolation warning: axis coordinate is out of limits (%lf; %lf) with value %lf, extrapolation is applied\n", axisMin, axisMax, axisCoordinate );
+      }
+    }
+
+    axisLow = axisIntervalIndex * axisStep + axisMin;
+    axisMult = (axisCoordinate - axisLow) * axisStepInv;
+    return axisIntervalIndex;
+  }
+
+
+  /**
+   * @brief interpolate all operator values and derivatives at a given point
+   * The algoritm is based on http://dx.doi.org/10.1090/S0025-5718-1988-0917826-0
+   *
+   * @param[in] axisCoordinates coordinates of a point
+   * @param[in] hypercubeData data of target hypercube
+   * @param[in] axisLows array of left coordinates of target axis intervals
+   * @param[in] axisMults array of weights of right coordinates of target axis intervals
+   * @param[in] axisStepInvs array of inversions of axis steps
+   * @param[out] values interpolated operator values
+   * @param[out] derivatives derivatives of interpolated operators
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  void
+  interpolatePointWithDerivatives( real64 const * const axisCoordinates,
+                                   real64 const * const hypercubeData,
+                                   real64 const * const axisLows,
+                                   real64 const * const axisMults,
+                                   real64 const * const axisStepInvs,
+                                   real64 *values,
+                                   real64 *derivatives ) const
```

Comment:
Thanks for the suggestion @rrsettgast! I've never used `restrict`, looks like it might actually help the compiler here.
This change was a bit late for this PR, so it will be included in the next one.
Will it be correct to use the type qualifier through `LVARRAY_RESTRICT` macros, and only for `values` and `derivatives` parameters?